### PR TITLE
Add initial golangci-lint action

### PR DESCRIPTION
**Pull request by cce**
## Summary

Run golangci-lint from Github actions using https://github.com/golangci/golangci-lint-action

## Test Plan

Adds new check to PRs and merges.

### CODE REVIEW

**Comment by winder** [comment]
Woah, this is really cool!

**Comment by cce** [comment]
Hrmm, too bad, reviewdog has not yet implemented the ability to pass through the different golangci-lint severity levels (errors vs warnings) to Github checks yet https://github.com/reviewdog/reviewdog/issues/343

File: .github/workflows/reviewdog.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+          echo "$GITHUB_WORKSPACE/bin" >> $GITHUB_PATH
+          echo "$RUNNER_WORKSPACE/$(basename $GITHUB_REPOSITORY)/bin" >> $GITHUB_PATH
+      - name: Install specific golang
+        uses: actions/setup-go@v2
+        with:
+          go-version: '1.16.6'
+      - name: Create folders for golangci-lint
+        run: mkdir -p cicdtmp/golangci-lint
+      - name: Check if custom golangci-lint is already built
+        id: cache-golnagci-lint
```
Comment (cce) [inline .github/workflows/reviewdog.yml:51]: typo here (and below)

File: .github/workflows/reviewdog.yml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+          --out-format line-number
+          -c .golangci-warnings.yml
+          --allow-parallel-runners
+          | reviewdog
+          -f=golangci-lint
+          -name="Lint Warnings"
+          -reporter=github-check
+          -filter-mode=added
+          -fail-on-error=false
+          -level=warning
```
Comment (cce) [inline .github/workflows/reviewdog.yml:90]: It seems to be showing up as red failure rather than a warning?

**Review by cce — COMMENTED** [review]
great!

File: .github/workflows/reviewdog.yml (hunk 2 above)
Comment (algojack) [reply .github/workflows/reviewdog.yml:90]: red warning, but non-blocking, i wasn't able to get it to show yellow if that's what you mean

File: .github/workflows/reviewdog.yml (hunk 1 above)
Comment (algojack) [reply .github/workflows/reviewdog.yml]: fixed ty

File: .golangci.yml [hunk 3]
```diff
@@ -1,35 +1,25 @@
...
-      original-url: github.com/algorand/go-algorand/cmd/partitiontest_linter
-#  govet:
-#    check-shadowing: true
+severity:
+  default-severity: error
 
 issues:
-  # don't use default exclude rules listed in `golangci-lint run --help`
+  # use these new lint checks on code since #2574
+  new-from-rev: 07d3c1a94a54a176a74c84ae762a41c746a93d3b
```
[the remark this answers is not in the document]
Comment (algojack) [reply .golangci.yml]: just did, ty

**Review by algobarb — COMMENTED** [review]
This is pretty cool.
So we're adding in the ReviewDog workflow which would act on github actions and it calls the tests directly, separating errors from warnings. So we moved partition-test away from `.golangci.yml` tests to `.golangci-warnings.yml` so that we only get warnings from reviewdogs instead of errors on the PRs right? What does changing `.golangci.yml` `tests` from `true` to `false` do?

File: .golangci-warnings.yml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      path: cmd/partitiontest_linter/plugin.so
+      description: This custom linter checks files that end in '_test.go', specifically functions that start with 'Test' and have testing argument, for a line 'partitiontest.ParitionTest(<testing arg>)'
+      original-url: github.com/algorand/go-algorand/cmd/partitiontest_linter
+
+severity:
+  default-severity: warning
+
+issues:
+  # use these new lint checks on code since #2574
+  new-from-rev: eb019291beed556ec6ac1ceb4a15114ce4df0c57
```
Comment (bricerisingalgorand) [inline .golangci-warnings.yml:30]: What happens when we add new linters, will we just update this line to the commit we add new linters?

File: .golangci-warnings.yml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+  # Disable default exclude rules listed in `golangci-lint run --help` (selectively re-enable some below)
+  exclude-use-default: false
+
+  # Maximum issues count per one linter. Set to 0 to disable. Default is 50.
+  max-issues-per-linter: 0
```
Comment (bricerisingalgorand) [inline .golangci-warnings.yml:36]: Very helpful comment, using 0 to disable this feels very counter intuitive lol

File: .golangci-warnings.yml (hunk 4 above)
Comment (cce) [reply .golangci-warnings.yml:30]: yes, or make an effort to "clean" out the repo for that specific linter like I did in https://github.com/algorand/go-algorand/pull/2574

**Comment by algojack** [comment]
> This is pretty cool.
> So we're adding in the ReviewDog workflow which would act on github actions and it calls the tests directly, separating errors from warnings. So we moved partition-test away from `.golangci.yml` tests to `.golangci-warnings.yml` so that we only get warnings from reviewdogs instead of errors on the PRs right? What does changing `.golangci.yml` `tests` from `true` to `false` do?

Tests: true -> false, tells the linters not to check test files that end in "_test.go". We have it true in warnings but false in blocking errors. And even in warnings we only limit them to partitiontest linter by excluding all other linters.

File: .golangci-warnings.yml (hunk 4 above)
Comment (algojack) [reply .golangci-warnings.yml:30]: yes? @cce what do you think?

**Review by algobarb — APPROVED** [review]
It looks good to me.

**Review by bricerisingalgorand — APPROVED** [review]

**Pull request merged**